source/fuzz/transformation_record_synonymous_constants.cpp

```diff
@@ -19,8 +19,61 @@ namespace spvtools {
 namespace fuzz {
 
 namespace {
-bool IsScalarZeroConstant(const opt::analysis::Constant* constant) {
-  return constant->AsScalarConstant() && constant->IsZero();
+
+// Returns true if the two given constants are equivalent
+// (the description of IsApplicable specifies the conditions they must satisfy
+// to be considered equivalent)
+bool AreEquivalentConstants(opt::IRContext* ir_context,
+                            const opt::analysis::Constant& constant1,
+                            const opt::analysis::Constant& constant2) {
+  // Check that the type ids are the same
+  // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/3536): Somehow
+  // relax this for integers (so that unsigned integer and signed integer are
+  // considered the same type)
+  uint32_t type_id1 = ir_context->get_type_mgr()->GetId(constant1.type());
+  uint32_t type_id2 = ir_context->get_type_mgr()->GetId(constant2.type());
+  if (type_id1 != type_id2) {
+    return false;
+  }
+
+  // If either constant is null, the other is equivalent iff it is zero-like
+  if (constant1.AsNullConstant()) {
+    return constant2.IsZero();
+  }
+
+  if (constant2.AsNullConstant()) {
+    return constant1.IsZero();
+  }
+
+  // If the constants are scalar, they are equal iff their words are the same
+  if (auto scalar1 = constant1.AsScalarConstant()) {
+    return scalar1->words() == constant2.AsScalarConstant()->words();
+  }
+
+  // If the constants are composite, they are equivalent iff their components
+  // match
+  if (constant1.AsCompositeConstant()) {
+    auto components1 = constant1.AsCompositeConstant()->GetComponents();
+    auto components2 = constant2.AsCompositeConstant()->GetComponents();
+
+    // Since the types match, we already know that the number of components is
+    // the same, so we just need to check equivalence for each one
+    for (size_t i = 0; i < components1.size(); i++) {
+      if (!AreEquivalentConstants(ir_context, *components1[i],
+                                  *components2[i])) {
+        return false;
+      }
+    }
```

Comment:
At this point, revert back to looking at the constant instructions and check whether their input operands (all of which must be constant ids) are pairwise equivalent.

Reply:
Done